pkg/common/cliprinter/protojson/protojson.go

```diff
@@ -0,0 +1,83 @@
+package protojson
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+
+	"github.com/spiffe/spire/pkg/common/cliprinter/errorjson"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+)
+
+// Print prints one or more protobuf messages formatted as JSON
+func Print(msgs []proto.Message, stdout, stderr io.Writer) {
+	if msgs == nil || len(msgs) == 0 {
+		return
+	}
+
+	jbs := [][]byte{}
+	jm := &protojson.MarshalOptions{
+		UseProtoNames: true,
+	}
+
+	// Unfortunately, we can only marshal one message at a time, so
+	// we need to build up an array of marshaled messages. We do this
+	// before printing them so we don't print an unterminated result
+	for _, msg := range msgs {
+		jb, err := jm.Marshal(msg)
+		if err != nil {
+			errorjson.Print(err, stdout, stderr)
+			return
+		}
+
+		// The golang protobuf encoders state in their documentation
+		// to not rely on the stability of their output. Furthermore,
+		// they deliberately inject randomness into the output to
+		// discourage reliance (and break our tests).
+		//
+		// Since we would like stable output, and we also want parity
+		// with the standard JSON encoding logic (and by extension
+		// our JSON output from non-proto sources), unmarshal then
+		// remarshal the output using the standard library.
+		//cj, err := launderJSON(jb)
+		//if err != nil {
+		//	errorjson.Print(err, stdout, stderr)
+		//	return
+		//}
+
+		//jbs = append(jbs, cj)
+		jbs = append(jbs, jb)
+	}
+
+	// Don't process this as a list if there's only one element
+	if len(jbs) == 1 {
+		fmt.Fprintln(stdout, string(jbs[0]))
+		return
+	}
+
+	start := "["
+	end := "]\n"
+	sep := ","
+
+	fmt.Fprint(stdout, start)
+	for i, jb := range jbs {
+		fmt.Fprint(stdout, string(jb))
+
+		// Don't print a separator after the last element
+		if i+1 < len(jbs) {
+			fmt.Fprint(stdout, sep)
+		}
+	}
+	fmt.Fprint(stdout, end)
```

Comment:
An alternative to manually constructing the formatting for the array would be to aggregate results into a []json.RawMessage slice and then use the standard json package to print out the results:

```
var jbs []json.RawMessage
.... aggregate stuffs...
json.NewEncoder(stdout).Encode(jbs)
```

Reply:
much better!